src/ocean/math/WideInt.d

```diff
@@ -0,0 +1,474 @@
+        test!("==")(num.payload[0], 0);
+        test!("==")(num.payload[1], 500);
+    }
+
+    /***************************************************************************
+
+        Mutates current number in-place, dividing it by `rhs` and calculating
+        remainder.
+
+        Params:
+            rhs = number to divide by
+
+        Returns:
+            remainder of division
+
+    ***************************************************************************/
+
+    public uint divideBy ( uint rhs )
+    {
+        ulong remainder = 0;
+
+        for (ptrdiff_t idx = this.payload.length - 1; idx >= 0; --idx)
+        {
+            remainder = (remainder << 32) + this.payload[idx];
+            ulong result = remainder / rhs;
+            remainder -= rhs * result;
+            assert(result <= uint.max);
+            this.payload[idx] = cast(uint) result;
+        }
+
+        assert(remainder <= uint.max);
+        return cast(uint) remainder;
+    }
+
+    unittest
+    {
+        WideInt num = 56 * 47;
+        test!("==")(0, num.divideBy(47));
+        test!("==")(num.payload[0], 56);
+
+        num.payload[0] = 1337;
+        num.payload[1] = 100;
+        test!("==")(1337, num.divideBy(1 << 31));
+        test!("==")(num.payload[0], 200);
+    }
+
+    /***************************************************************************
+
+        Mutates this number in place multiplying it by a regular integer
+
+        Params:
+            rhs = value to multiply by
+
+        Throws:
+            WideIntRangeException if resulting value can't fit in current
+            WideInt value range
+
+    ***************************************************************************/
+
+    public void multiplyBy ( uint rhs )
+    {
+        ulong overflow = 0;
+
+        for (size_t i = 0; i < this.payload.length; ++i)
+        {
+            overflow += cast(ulong)(this.payload[i]) * rhs;
+            this.payload[i] = cast(uint) (overflow & uint.max);
+            overflow >>= 32;
+        }
+
+        enforce(.wideint_exception, overflow == 0);
+    }
+
+    unittest
+    {
+        WideInt num = 1_000_000_000;
+        num.multiplyBy(1_000_000_000);
+        test!("==")(num.toString(), "1000000000000000000");
+    }
+
+    /***************************************************************************
+
+        Returns:
+            double precision floating point representation of this number
+
+    ***************************************************************************/
+
+    public double toDouble ( )
+    {
+        // TODO: for widenumbers with value > ulong.max need to calculate
+        // mantiss and exponent manually by counting bitshift
+
+        ulong last = this.payload[1];
+        last <<= 32;
+        last += this.payload[0];
+        return cast(double) last;
+    }
+
+    unittest
+    {
+        WideInt num = 1_222_333_444;
+        num.multiplyBy(1_000_000_000);
+        test!("is")(num.toDouble(), 1_222_333_444e+9);
```

Comment:
Do you know that for sure? 1,222,333,444,000,000,000 is divisible by 2048 so this test will still pass if you lose 11 mantissa bits and may fail for 1,222,333,444,000,000,001 or even 1,222,333,444,100,000,000.
Ok, you *are* right, but maybe verify this by something like
```d
static assert(1_222_333_444_000_000_000 > (1 << double.mant_dig));
```